packages/core/src/auth/tokens/token-scope.ts

```diff
@@ -1,26 +1,66 @@
 export enum CacheRole {
   None = 'none',
   ReadWrite = 'readwrite',
+  ReadOnly = 'readonly',
 }
 
+export class All {}
+export class CacheName {
+  name: string;
+  constructor(name: string) {
+    this.name = name;
+  }
+}
+export type CacheResource = All | CacheName;
+
 export class CachePermission {
   cacheRole: CacheRole;
+  cache: CacheResource;
 
-  constructor(cacheRole: CacheRole) {
+  constructor(cacheRole: CacheRole, cache?: CacheResource) {
     this.cacheRole = cacheRole;
+    if (cache) {
+      this.cache = cache;
+    } else {
+      this.cache = new All();
+    }
   }
 }
 
 export enum TopicRole {
   None = 'none',
   ReadWrite = 'readwrite',
+  ReadOnly = 'readonly',
 }
 
+export class TopicName {
+  name: string;
+  constructor(name: string) {
+    this.name = name;
+  }
+}
+export type TopicResource = All | TopicName;
+
 export class TopicPermission {
   topicRole: TopicRole;
-
-  constructor(topicRole: TopicRole) {
+  cache: CacheResource;
+  topic: TopicResource;
+  constructor(
+    topicRole: TopicRole,
+    cache?: CacheResource,
+    topic?: TopicResource
```

Comment:
when you start having multiple optional positional args in JS it is common to create an interface to represent them. e.g.:

```
export interface TopicPermissionOpts {
   cache?: CacheResource,
   topic?: TopicResource
}
```

Then users can call it like this:

```
new TopicPermission(role, {topic: ...});
```

Otherwise it will feel weird that you are allowing `cache` to be optional but basically forcing them to pass `undefined` explicitly if they don't want to set it.